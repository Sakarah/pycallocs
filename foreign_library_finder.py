import allocs
import importlib
import importlib.abc
import importlib.machinery
import sys

# These can be modified by the user to change the finder search paths
lib_extension = ".so"

# The blank search path has a special meaning as dlopen is using its own search 
# paths when the file name does not contain any /
search_paths = [""]

class ForeignLibraryFinder(importlib.abc.MetaPathFinder):
    """
        Meta path finder searching foreign libraries usable
        with allocs.ForeignLibraryLoader.
    """

    def find_spec(fullname, path, target):
        if path is not None:
            # Should we support submodules ?
            return None
        
        for base_path in search_paths:
            try:
                filename = base_path + fullname + lib_extension
                loader = allocs.ForeignLibraryLoader(filename)
                return importlib.machinery.ModuleSpec(fullname, loader, origin=filename)
            except ImportError:
                continue

        return None

sys.meta_path.append(ForeignLibraryFinder)
