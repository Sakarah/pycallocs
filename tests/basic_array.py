import foreign_library_finder
foreign_library_finder.search_paths.append("libs/")
import basic as m

int_arr = m.uint_16.array([1,2,3,4,5])
print(int_arr)

char_arr = m.char.array("why not ?")
print(char_arr)
byte_recons = bytes()
for elt in char_arr:
    byte_recons += elt
print(byte_recons)

