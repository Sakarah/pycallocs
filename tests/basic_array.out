<[1, 2, 3, 4, 5]>
<[b'w', b'h', b'y', b' ', b'n', b'o', b't', b' ', b'?']>
b'why not ?'
